lib/iris/tests/_shared_utils.py

```diff
@@ -0,0 +1,929 @@
+                "\n  with shapes={} {}, atol={}, rtol={}"
+                "\n  worst at element {} :  a={}  b={}"
+                "\n  absolute error ~{:.3g}, equivalent to rtol ~{:.3e}"
+            )
+            aval, bval = a[worst_inds], b[worst_inds]
+            absdiff = np.abs(aval - bval)
+            equiv_rtol = absdiff / bval
+            msg = msg.format(
+                a.shape,
+                b.shape,
+                atol,
+                rtol,
+                worst_inds,
+                aval,
+                bval,
+                absdiff,
+                equiv_rtol,
+            )
+
+        raise AssertionError(msg)
+
+
+def file_checksum(file_path):
+    """Generate checksum from file."""
+    with open(file_path, "rb") as in_file:
+        return zlib.crc32(in_file.read())
+
+
+def _check_reference_file(reference_path):
+    reference_exists = os.path.isfile(reference_path)
+    if not (reference_exists or os.environ.get("IRIS_TEST_CREATE_MISSING")):
+        msg = "Missing test result: {}".format(reference_path)
+        raise AssertionError(msg)
+    return reference_exists
+
+
+def _ensure_folder(path):
+    dir_path = os.path.dirname(path)
+    if not os.path.exists(dir_path):
+        os.makedirs(dir_path)
+
+
+# todo: need to find equlivalence for `unique_id` in pytest
+def check_graphic():
+    """Check the hash of the current matplotlib figure matches the expected
+    image hash for the current graphic test.
+
+    To create missing image test results, set the IRIS_TEST_CREATE_MISSING
+    environment variable before running the tests. This will result in new
+    and appropriately "<hash>.png" image files being generated in the image
+    output directory, and the imagerepo.json file being updated.
+
+    """
+    assert False
+
+
+# todo: relied on unitest functionality, need to find a pytest alternative
+def patch(*args, **kwargs):
+    """Install a mock.patch, to be removed after the current test.
+
+    The patch is created with mock.patch(*args, **kwargs).
+
+    Returns
+    -------
+    The substitute object returned by patch.start().
+
+    Examples
+    --------
+    ::
+
+        mock_call = self.patch('module.Class.call', return_value=1)
+        module_Class_instance.call(3, 4)
+        self.assertEqual(mock_call.call_args_list, [mock.call(3, 4)])
+
+    """
+    raise NotImplementedError()
+
+
+def assert_array_shape_stats(result, shape, mean, std_dev, rtol=1e-6):
+    """Assert that the result, a cube, has the provided shape and that the
+    mean and standard deviation of the data array are also as provided.
+    Thus build confidence that a cube processing operation, such as a
+    cube.regrid, has maintained its behaviour.
+
+    """
+    assert result.shape == shape
+    assert_array_all_close(result.data.mean(), mean, rtol=rtol)
+    assert_array_all_close(result.data.std(), std_dev, rtol=rtol)
+
+
+def assert_dict_equal(lhs, rhs, msg=None):
+    """Dictionary Comparison."""
```

Comment:
The original docstring explained why this was necessary. It would be good to restore that.